File: DSA-for-Beginners/01_Array/02_Dynamic_Array.py
# Python arrays are dynamic by default, but this is an example of resizing
class DynamicArray:
    def __init__(self, capacity):
        self.capacity = capacity
        self.length = 0
        self.arr = [0] * capacity
    
    # Insert val in the last position of the array
    def pushback(self, val):
        if self.length == self.capacity:
            self.resize()

        # Insert at next empty position
        self.arr[self.length] = val
        self.length += 1
    
    def resize(self):
        # Create new array of double capacity
        self.capacity = 2 * self.capacity
        newArr = [0] * self.capacity
        
        # Copy elements to newArr
        for i in range(self.length):
            newArr[i] = self.arr[i]
        
        self.arr = newArr
    
    # Remove the last element in the array
    def popback(self):
        if self.length > 0:
            self.length -= 1

    # Get value at i-th index
    def get(self, i):
        if i < self.length:
            return self.arr[i]
        # Here we would throw an out of bounds exception

    # Insert val at i-th index
    def set(self, i, val):
        if i < self.length:
            self.arr[i] = val
        # Here we would throw and out of bounds exception

    # Return number of elements in the dynamic array
    def getSize(self):
        return self.length

    # Return capacity of the dynamic array
    def getCapacity(self):
        return self.capacity

def main():
    # Test Case 1
    # Input: ["Array", 1, "getSize", "getCapacity"]
    print("Test Case #1")
    arr1 = DynamicArray(1)
    print(f"getSize: {arr1.getSize()}")
    print(f"getCapacity: {arr1.getCapacity()}")
    print("\n")
    
    # Test Case 2
    # Input: ["Array", 1, "pushback", 1, "getCapacity", "pushback", 2, "getCapacity"]
    print("Test Case #2")
    arr2 = DynamicArray(1)
    arr2.pushback(1)
    print(f"getCapacity: {arr2.getCapacity()}")
    arr2.pushback(2)
    print(f"getCapacity: {arr2.getCapacity()}")
    print("\n")

    # Test Case 3
    # Input: ["Array", 1, "getSize", "getCapacity", "pushback", 1, "getSize", "getCapacity", 
    #         "pushback", 2, "getSize", "getCapacity", "get", 1, "set", 1, 3, "get", 1, 
    #         "popback", "getSize", "getCapacity"]
    print("Test Case #3")
    arr3 = DynamicArray(1)
    print(f"getSize: {arr3.getSize()}")
    print(f"getCapacity: {arr3.getCapacity()}")
    arr3.pushback(1)
    print(f"getSize: {arr3.getSize()}")
    print(f"getCapacity: {arr3.getCapacity()}")
    arr3.pushback(2)
    print(f"getSize: {arr3.getSize()}")
    print(f"getCapacity: {arr3.getCapacity()}")
    arr3.get(1)
    arr3.set(1, 3)
    arr3.get(1)
    arr3.popback()
    print(f"getSize: {arr3.getSize()}")
    print(f"getCapacity: {arr3.getCapacity()}")
    print("\n")
    
if __name__ == "__main__":
    main()

File: DSA-for-Beginners/01_Array/02_Dynamic_Array.cpp
#include <iostream>

class DynamicArray {
private:
    int capacity = 0;
    int length = 0;
    int* arr;
public:
    // Constructor
    DynamicArray(int capacity) {
        this->capacity = capacity;
        arr = new int[capacity];
    }

    // Return number of elements in the dynamic array
    int getSize() {
        return length;
    }
    
    // Return capacity of the dynamic array
    int getCapacity() {
        return capacity;
    }

    // Get value at i-th index
    int get(int i) {
        return arr[i];
    }

    // Insert val at i-th index
    void set(int i, int val) {
        arr[i] = val;
    }

    // Insert val in the last position of the array
    void pushback(int val) {
        if(length == capacity) {
            resize();
        }
        arr[length] = val;
        length++;
    }

    // Remove the last element in the array
    int popback() {
        length--;
        return arr[length];
    }

    // Resize the array by doubling the capacity
    void resize() {
        capacity = 2 * capacity;
        int* newArr = new int[capacity];
        
        // Copy elements to newArr
        for (int i = 0; i < length; i++) {
            newArr[i] = arr[i];
        }

        delete[] arr;
        arr = newArr;
    }
};

int main() {
    // Test Case #1
    // Input: ["Array", 1, "getSize", "getCapacity"]
    printf("Test Case #1\n");
    DynamicArray Array1(1);
    printf("getSize: %d\n", Array1.getSize());
    printf("getCapacity: %d\n\n", Array1.getCapacity());

    // Test Case #2
    printf("Test Case #2\n");
    // Input: ["Array", 1, "pushback", 1, "getCapacity", "pushback", 2, "getCapacity"]
    DynamicArray Array2(1);
    Array2.pushback(1);
    printf("getCapacity: %d\n", Array2.getCapacity());
    Array2.pushback(2);
    printf("getCapacity: %d\n\n", Array2.getCapacity());
    
    // Test Case #3
    printf("Test Case #3\n");
    // Input: ["Array", 1, "getSize", "getCapacity", "pushback", 1, "getSize", "getCapacity", 
    //         "pushback", 2, "getSize", "getCapacity", "get", 1, "set", 1, 3, "get", 1, 
    //         "popback", "getSize", "getCapacity"]
    DynamicArray Array3(1);
    printf("getSize: %d\n", Array3.getSize());
    printf("getCapacity: %d\n", Array3.getCapacity());
    Array3.pushback(1);
    printf("getSize: %d\n", Array3.getSize());
    printf("getCapacity: %d\n", Array3.getCapacity());
    Array3.pushback(2);
    printf("getSize: %d\n", Array3.getSize());
    printf("getCapacity: %d\n", Array3.getCapacity());
    Array3.get(1);
    Array3.set(1, 3);
    Array3.get(1);
    Array3.popback();
    printf("getSize: %d\n", Array3.getSize());
    printf("getCapacity: %d\n\n", Array3.getCapacity());

    return 0;
}